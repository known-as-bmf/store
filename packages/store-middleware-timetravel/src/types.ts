export interface TimetravelOptions {
  /**
   * The depth of the history.
   * Must be a positive number greater or equal to 1.
   * @default 1
   */
  depth?: number;
}
