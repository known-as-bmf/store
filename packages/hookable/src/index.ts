export { createHookable } from './hookable';
export { every, once } from './hook-subscribers';
export {
  Hook,
  HookSubscriber,
  Hooks,
  TapInput,
  TapOutput,
  TransformInput,
  TransformOutput,
} from './types';
