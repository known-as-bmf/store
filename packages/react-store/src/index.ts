export { useStore } from './useStore';
export { useNewStore } from './useNewStore';
