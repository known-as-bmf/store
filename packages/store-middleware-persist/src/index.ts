export { PersistOptions } from './types';
export { persistMiddleware } from './persist';
