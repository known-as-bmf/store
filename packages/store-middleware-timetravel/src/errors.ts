export const errors = {
  invalidStore:
    'Provided value is not a valid store instance or has not been initialized with timetravel middleware.',
  historyDepthValue: (depth: number): string =>
    `historyDepth must be greater than 0, got ${depth}`,
  negativeStep: (steps: number): string =>
    `steps must be greater than or equal to 0, got ${steps}`,
  outOfBoundsBackward: 'Not enough states in history stack to go backward.',
  outOfBoundsForward: 'Not enough states in history stack to go forward.',
};
