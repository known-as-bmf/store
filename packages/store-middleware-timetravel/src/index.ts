export { TimetravelOptions } from './types';
export { timetravelMiddleware, undo, redo, derefHistory } from './timetravel';
