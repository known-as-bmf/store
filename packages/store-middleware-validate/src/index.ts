export { Validator } from './types';
export { validatorMiddleware } from './validator';
