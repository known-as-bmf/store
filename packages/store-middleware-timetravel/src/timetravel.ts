import { Store, Middleware, set } from '@known-as-bmf/store';

import { TimetravelOptions } from './types';
import { errors } from './errors';

const HISTORY = Symbol();
const HISTORY_INDEX = Symbol();
const IS_TIMETRAVELING = Symbol();

interface InternalStore<S> extends Store<S> {
  [HISTORY]: S[];
  [HISTORY_INDEX]: number;
  [IS_TIMETRAVELING]: boolean;
}

function assertTimetravelStore<S>(
  store: Store<S>
): asserts store is InternalStore<S> {
  if (!Object.prototype.hasOwnProperty.call(store, HISTORY)) {
    throw new TypeError(errors.invalidStore);
  }
}

function navigateHistory<S>(store: InternalStore<S>, offset: number): void {
  if (offset === 0) {
    // nothing to do boys
    return;
  }

  const { [HISTORY]: history, [HISTORY_INDEX]: index } = store;

  const futureIndex = index + offset;

  if (futureIndex >= history.length) {
    throw new Error(errors.outOfBoundsBackward);
  }

  if (futureIndex < 0) {
    throw new Error(errors.outOfBoundsForward);
  }

  store[HISTORY_INDEX] = futureIndex;
  store[IS_TIMETRAVELING] = true;
  set(store, store[HISTORY][futureIndex]);
}

export const timetravelMiddleware = <S>({
  depth = 1,
}: TimetravelOptions = {}): Middleware<S> => (store, hooks) => {
  if (depth < 1) {
    throw new Error(errors.historyDepthValue(depth));
  }

  const enhancedStore = Object.defineProperties(store, {
    [HISTORY]: { value: [], writable: true },
    [HISTORY_INDEX]: { value: 0, writable: true },
    [IS_TIMETRAVELING]: { value: false, writable: true },
  }) as InternalStore<S>;

  hooks.stateDidChange(() => (state) => {
    const {
      [HISTORY]: history,
      [HISTORY_INDEX]: index,
      [IS_TIMETRAVELING]: isTimetraveling,
    } = enhancedStore;

    if (isTimetraveling) {
      enhancedStore[IS_TIMETRAVELING] = false;
    } else {
      // do some history stuff
      // skip forward history, prepend new state, take n elements (history depth)
      enhancedStore[HISTORY] = [state, ...history.slice(index)].slice(0, depth);
      enhancedStore[HISTORY_INDEX] = 0;
    }
  });
};

/**
 * Returns the current state of a store and the available history (past and future).
 * @param store The store you want to get the current state from.
 * @throws {TypeError} if the store is not a correct `Store` instance or if you didn't pass the timetravel middleware during store construction.
 */
export function derefHistory<S>(
  store: Store<S>
): { future: S[]; past: S[]; current: S } {
  assertTimetravelStore(store);

  const { [HISTORY]: history, [HISTORY_INDEX]: index } = store;
  return {
    future: history.slice(0, index),
    past: history.slice(index + 1),
    current: history[index],
  };
}

/**
 * Navigates back in the store history.
 * @param store The store of which you want to change the state.
 * @param steps The number of steps you want to take backwards. Default to 1.
 * @throws {TypeError} if the store is not a correct `Store` instance or if you didn't pass the timetravel middleware during store construction.
 * @throws {Error} if `steps` is less than 0.
 * @throws {Error} if you try to navigate out of bounds of the history.
 */
export function undo<S>(store: Store<S>, steps = 1): void {
  assertTimetravelStore(store);

  if (steps < 0) {
    throw new Error(errors.negativeStep(steps));
  }

  navigateHistory(store, steps);
}

/**
 * Navigates forward in the store history.
 * @param store The store of which you want to change the state.
 * @param steps The number of steps you want to take forward. Default to 1.
 * @throws {TypeError} if the store is not a correct `Store` instance or if you didn't pass the timetravel middleware during store construction.
 * @throws {Error} if `steps` is less than 0.
 * @throws {Error} if you try to navigate out of bounds of the history.
 */
export function redo<S>(store: Store<S>, steps = 1): void {
  assertTimetravelStore(store);

  if (steps < 0) {
    throw new Error(errors.negativeStep(steps));
  }

  navigateHistory(store, -steps);
}
