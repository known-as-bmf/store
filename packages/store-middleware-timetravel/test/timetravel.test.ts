import { of, swap, set, deref, subscribe } from '@known-as-bmf/store';

import { timetravelMiddleware, derefHistory, redo, undo } from '../src';

describe('timetravelMiddleware', () => {
  it('should be a function', () => {
    expect(timetravelMiddleware).toBeDefined();
    expect(timetravelMiddleware).toBeInstanceOf(Function);
  });
});

describe('of', () => {
  it('should properly initialize the history', () => {
    const store = of('a', timetravelMiddleware());

    expect(derefHistory(store)).toMatchSnapshot();
  });

  it.each([-2, -1, -0, 0, +0])(
    'should throw error if options are invalid',
    (depth) => {
      expect(() =>
        of(1, timetravelMiddleware({ depth }))
      ).toThrowErrorMatchingSnapshot();
    }
  );
});

describe('derefHistory', () => {
  it('should be a function', () => {
    expect(derefHistory).toBeDefined();
    expect(derefHistory).toBeInstanceOf(Function);
  });

  it('should return an object with 3 keys (future, past, current)', () => {
    const a = 1;
    const b = 2;
    const c = 3;

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);
    set(store, c);

    undo(store);

    expect(derefHistory(store)).toMatchSnapshot();
  });

  it('should throw if not given a store', () => {
    expect(() => derefHistory({})).toThrowErrorMatchingSnapshot();
    expect(() => derefHistory({ ...of(1) })).toThrowErrorMatchingSnapshot();
  });
});

describe('swap', () => {
  it('should properly update the history', () => {
    const store = of<string>('a', timetravelMiddleware({ depth: 5 }));

    swap(store, () => 'b');

    expect(derefHistory(store)).toMatchSnapshot();
  });
});

describe('set', () => {
  it('should properly update the history', () => {
    const store = of<string>('a', timetravelMiddleware({ depth: 5 }));

    set(store, 'b');

    expect(derefHistory(store)).toMatchSnapshot();
  });
});

describe('undo', () => {
  it('should be a function', () => {
    expect(undo).toBeDefined();
    expect(undo).toBeInstanceOf(Function);
  });

  it('should go backward 1 step if no steps specified', () => {
    const a = {};
    const b = {};

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);

    undo(store);

    expect(deref(store)).toBe(a);
  });

  it('should go backward n step if steps specified', () => {
    const a = {};
    const b = {};
    const c = {};

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);
    set(store, c);

    undo(store, 2);

    expect(deref(store)).toBe(a);
  });

  it('should do nothing if steps is 0', () => {
    const a = {};
    const b = {};
    const c = {};
    const callback = jest.fn();

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);
    set(store, c);

    subscribe(store, callback);

    undo(store, 0);

    expect(deref(store)).toBe(c);
    expect(callback).not.toHaveBeenCalled();
  });

  it('should notify subscribers with new state', () => {
    const a = 1;
    const b = 2;
    const callback = jest.fn();

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);

    subscribe(store, callback);

    undo(store);

    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith({ previous: b, current: a });
  });

  it('should throw if given a negative step value', () => {
    expect(() =>
      undo(of(1, timetravelMiddleware()), -2)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if trying to go backward when there is no past', () => {
    expect(() =>
      undo(of(1, timetravelMiddleware()))
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if not given a store', () => {
    expect(() => undo({})).toThrowErrorMatchingSnapshot();
    expect(() => undo({ ...of(1) })).toThrowErrorMatchingSnapshot();
  });
});

describe('redo', () => {
  it('should be a function', () => {
    expect(redo).toBeDefined();
    expect(redo).toBeInstanceOf(Function);
  });

  it('should go forward 1 step if no steps specified', () => {
    const a = {};
    const b = {};

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);

    undo(store);
    redo(store);

    expect(deref(store)).toBe(b);
  });

  it('should go back n step if steps specified', () => {
    const a = {};
    const b = {};
    const c = {};

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);
    set(store, c);

    undo(store, 2);
    redo(store, 2);

    expect(deref(store)).toBe(c);
  });

  it('should do nothing if steps is 0', () => {
    const a = {};
    const b = {};
    const c = {};
    const callback = jest.fn();

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);
    set(store, c);

    subscribe(store, callback);

    redo(store, 0);

    expect(deref(store)).toBe(c);
    expect(callback).not.toHaveBeenCalled();
  });

  it('should notify subscribers with new state', () => {
    const a = 1;
    const b = 2;
    const callback = jest.fn();

    const store = of(a, timetravelMiddleware({ depth: 5 }));

    set(store, b);
    undo(store);

    subscribe(store, callback);

    redo(store);

    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith({ previous: a, current: b });
  });

  it('should throw if given a negative step value', () => {
    expect(() =>
      redo(of(1, timetravelMiddleware()), -2)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if trying to go forward when there is no future', () => {
    expect(() =>
      redo(of(1, timetravelMiddleware()))
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if not given a store', () => {
    expect(() => redo({})).toThrowErrorMatchingSnapshot();
    expect(() => redo({ ...of(1) })).toThrowErrorMatchingSnapshot();
  });
});
