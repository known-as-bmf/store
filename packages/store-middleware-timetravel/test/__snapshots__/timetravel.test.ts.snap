// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`derefHistory should return an object with 3 keys (future, past, current) 1`] = `
Object {
  "current": 2,
  "future": Array [
    3,
  ],
  "past": Array [
    1,
  ],
}
`;

exports[`derefHistory should throw if not given a store 1`] = `"Provided value is not a valid store instance or has not been initialized with timetravel middleware."`;

exports[`derefHistory should throw if not given a store 2`] = `"Provided value is not a valid store instance or has not been initialized with timetravel middleware."`;

exports[`of should properly initialize the history 1`] = `
Object {
  "current": "a",
  "future": Array [],
  "past": Array [],
}
`;

exports[`of should throw error if options are invalid 1`] = `"historyDepth must be greater than 0, got -2"`;

exports[`of should throw error if options are invalid 2`] = `"historyDepth must be greater than 0, got -1"`;

exports[`of should throw error if options are invalid 3`] = `"historyDepth must be greater than 0, got 0"`;

exports[`of should throw error if options are invalid 4`] = `"historyDepth must be greater than 0, got 0"`;

exports[`of should throw error if options are invalid 5`] = `"historyDepth must be greater than 0, got 0"`;

exports[`redo should throw if given a negative step value 1`] = `"steps must be greater than or equal to 0, got -2"`;

exports[`redo should throw if not given a store 1`] = `"Provided value is not a valid store instance or has not been initialized with timetravel middleware."`;

exports[`redo should throw if not given a store 2`] = `"Provided value is not a valid store instance or has not been initialized with timetravel middleware."`;

exports[`redo should throw if trying to go forward when there is no future 1`] = `"Not enough states in history stack to go forward."`;

exports[`set should properly update the history 1`] = `
Object {
  "current": "b",
  "future": Array [],
  "past": Array [
    "a",
  ],
}
`;

exports[`swap should properly update the history 1`] = `
Object {
  "current": "b",
  "future": Array [],
  "past": Array [
    "a",
  ],
}
`;

exports[`undo should throw if given a negative step value 1`] = `"steps must be greater than or equal to 0, got -2"`;

exports[`undo should throw if not given a store 1`] = `"Provided value is not a valid store instance or has not been initialized with timetravel middleware."`;

exports[`undo should throw if not given a store 2`] = `"Provided value is not a valid store instance or has not been initialized with timetravel middleware."`;

exports[`undo should throw if trying to go backward when there is no past 1`] = `"Not enough states in history stack to go backward."`;
